User Export

dgitis opened the pull request:
## Description & motivation
Work-in-progress resolving #285.

The goal of this PR is to support the `pseudonymous_user_id` and `user_id` tables.

My initial thought is that we should keep both the package's current user tables and new ones derived from the new user export options in GA4. The reason for this is to support both old and new installations.

README modifications are not done, but I expect we will want to add a new section on disabling user models that explains the differences between our various models and how to use them.

What defaults should we have for `+enable` on these tables?

This PR supports `audiences` fields from the export. While working with a test site that has implemented audiences, the data doesn't seem to be very useful unless its ID field joins with Google Ads audience exports. Despite this, I think we should leave this in because it needs to be enabled by a variable.

The `user_properties` fields were done without sample data. Naming and logic will likely need to be updated.

Multi-site is not currently supported.

We also might want to move the logic in `base_ga4___*` models to `stg_ga4__*` models.

I also need to review package naming conventions to ensure data, like geo data, is consistent with elsewhere in the package.

To-do:
- data marts
- review package naming conventions to ensure consistency
- `user_property` logic
- readme
- multi-site

Here are the [docs for the source tables](https://support.google.com/analytics/answer/12769371?hl=en&ref_topic=9359001&sjid=5665659258369477-NC).

## Checklist
- [ ] I have verified that these changes work locally
- [ ] I have updated the README.md (if applicable)
- [ ] I have added tests & descriptions to my models (and macros if applicable)
- [ ] I have run `dbt test` and `python -m pytest .` to validate existing tests

dgitis commented:
@adamribaudo-velir, while I've verified that these work locally, I've only verified with a minimal amount of data. Multi-site changes were verified by setting the same property ID several times under `property_ids`. Obviously this is not ideal.

It would be best to verify this using a proper multi-site installation. I don't have any current clients, but I can ask some past clients.

I did not write any tests. [dbt recommends testing certain](https://docs.getdbt.com/docs/build/unit-tests#when-to-add-a-unit-test-to-your-model) things. The only logic worth testing , the multi-site macro, is not testable by the new framework.

I defaulted these new tables to `+enable: false`. 

I also ran `base_ga4__pseudonymous_users` in to `stg_ga4__client_ids`. I renamed at the staging model because that's where the client IDs are generated.

adamribaudo-velir commented on macros/combine_property_data.sql:
```diff
@@ -6,41 +6,69 @@
     {% if not should_full_refresh() %}
         {# If incremental, then use static_incremental_days variable to find earliest shard to copy #}
         {%- set earliest_shard_to_retrieve = (modules.datetime.date.today() - modules.datetime.timedelta(days=var('static_incremental_days')))|string|replace("-", "")|int -%}
```
@dgitis I think you've introduced some duplicate code in this macro?

adamribaudo-velir commented:
I'm on board with the approach, but unable to run it due to this error. Couldn't find an obvious issue.  I can run the main branch fine with the same project configuration. 

```
20:08:18  Completed with 1 error and 0 warnings:
20:08:18
20:08:18    Compilation Error in model base_ga4__events (models\staging\base\base_ga4__events.sql)
  can only concatenate str (not "int") to str

  > in macro combine_property_data (macros\combine_property_data.sql)
  > called by macro run_hooks (macros\materializations\hooks.sql)
  > called by macro materialization_incremental_bigquery (macros\materializations\incremental.sql)
  > called by model base_ga4__events (models\staging\base\base_ga4__events.sql)
```

dgitis commented:
I fixed the duplicate code issues which seems to have fixed the error, not sure why though, and also updated a test that worked on my very small test site but wouldn't work on a larger site.
